File: QuickSort/quick.lua
function quickSort(arr, p, r)
	if p < r then
		local q = partition(arr, p, r)
		quickSort(arr, p, q - 1)
		quickSort(arr, q + 1, r)
	end
end

function partition(arr, p, r)
	x = arr[r]
	i = p - 1
	for j = p, r - 1 do
		if arr[j] <= x then
			i = i + 1
			arr[i], arr[j] = arr[j], arr[i]
		end
	end
	arr[i+1], arr[r] = arr[r], arr[i+1]
	return i + 1
end

function main()
	io.write("Enter number of elements: ")
	n = io.read("*n")
	arr = {}

	for i = 1, n do
		io.write("Enter an element: ")
		arr[i] = io.read("*n")
	end

	p, r = 1, n

	quickSort(arr, p, r)
	
	print("Sorted array is: ")
	for i = 1, n do
		io.write(arr[i].." ")
	end
end

main()

File: QuickSort/quick.cpp
#include <iostream>

#define cin std::cin
#define cout std::cout
#define endl std::endl

void swap(int &a, int &b) {
	int temp = a;
	a = b;
	b = temp;
}

int partition(int arr[], int p, int r) {
	int x = arr[r];
	int i = p - 1;
	for(int j = p; j <= r - 1; j++) {
		if(arr[j] <= x) {
			i++;
			swap(arr[i], arr[j]);
		}
	}
	swap(arr[i + 1], arr[r]);
	return i + 1;
}

void quickSort(int arr[], int p, int r) {
	if(p < r) {
		int q = partition(arr, p, r);
		quickSort(arr, p, q-1);
		quickSort(arr, q+1, r);
	}
}

int main() {
	int n;
	cout << "Enter number of elements: ";
	cin >> n;
	int arr[n], p = 0, r = n - 1;
	for(int i = 0; i < n; i++) {
		cout << "Enter an element: ";
		cin >> arr[i];
	} 

	quickSort(arr, p, r);

	cout << "The sorted array is: " << endl;
	for(int i = 0; i < n; i++) {
		cout << arr[i] << " ";
	}
}
